### PR TITLE
Upgrade pulumi-aws to v6 and pulumi-kubernetes to v4

### LINKED ISSUE #902: [Epic] Pulumi-EKS: Update pulumi-aws to v6
Update pulumi-aws to V6

```[tasklist]
### Tasks
- [ ] [Update internal dependencies and examples within provider](https://github.com/pulumi/pulumi-eks/pull/910)
- [ ] [Add upgrade tests](https://github.com/pulumi/pulumi-eks/issues/927)
- [ ] https://github.com/pulumi/examples/issues/1517
- [ ] [Update pulumi-templates](https://github.com/pulumi/templates/pull/650)
```

### LINKED ISSUE #908: Upgrade to pulumi-kubernetes v4
Ideally, this upgrade could be combined with #902  into a new major version.

### LINKED ISSUE #926: NodeGroupV2 not encrypting block storage
### What happened?

When disabling the default NodeGroup in eks.Cluster and enabling `encryptRootBlockDevice: true` for eks.NodeGroupV2, the launch template in use by the Autoscaling group is not setting the storage to be encrypted.

### Example

```typescript
export const eksCluster = new eks.Cluster(resourcePrefix + "eks-clstr", {
  name: resourcePrefix + "eks-clstr",
  vpcId: vpcId,
  clusterSecurityGroup: sgEks,
  privateSubnetIds: eksSubnetsIds,
  nodeAssociatePublicIpAddress: false,
  instanceType: cfgEks.require("node-size"),
  version: cfgEks.require("version"),
  nodeAmiId: eksAmiId,
  desiredCapacity: cfgEks.requireNumber("node-count"),
  minSize: 4,
  maxSize: 9,
  skipDefaultNodeGroup: true,
  serviceRole: eksSvcRole,
  instanceRole: eksNodeRole,
  encryptRootBlockDevice: true,
  endpointPublicAccess: false,
  endpointPrivateAccess: true,
  createOidcProvider: true,
  tags: {
    Name: resourcePrefix + "eks-clstr",
  },
});

new eks.NodeGroupV2(resourcePrefix + "NodeGroup", {
  cluster: eksCluster,
  instanceType: cfgEks.require("node-size"),
  desiredCapacity: cfgEks.requireNumber("node-count"),
  amiId: eksAmiId,
  nodeSubnetIds: eksCluster.core.subnetIds,
  extraNodeSecurityGroups: [sgEks],
  nodeAssociatePublicIpAddress: false,
  minSize: 4,
  maxSize: 9,
  encryptRootBlockDevice: true,
  instanceProfile: eksInstanceProfile,
});
```

### Output of `pulumi about`

❯ pulumi about
CLI          
Version      3.88.1
Go Version   go1.21.1
Go Compiler  gc

Plugins
NAME                  VERSION
aws                   5.41.0
aws                   5.31.0
eks                   1.0.2
kubernetes            3.29.0
nodejs                unknown
random                4.13.2
tls                   4.10.0
tls-self-signed-cert  0.1.3

Host     
OS       darwin
Version  13.5.1
Arch     x86_64

This project is written in nodejs: executable='/Users/username/brew/bin/node' version='v20.8.1'

Dependencies:
NAME                          VERSION
@types/node                   16.18.34
typescript                    5.1.3
@pulumi/aws                   5.41.0
@pulumi/eks                   1.0.2
@pulumi/pulumi                3.78.1
@pulumi/random                4.13.2
@pulumi/tls-self-signed-cert  0.1.3
@pulumi/tls                   4.10.0

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### LINKED ISSUE #930: Docs for nodegroup and nodegroupV2 show deprecated root block encryption property.
### What happened?

The docs for the package show this property:
[encryptRootBlockDevice](https://www.pulumi.com/registry/packages/eks/api-docs/nodegroupv2/#encryptrootblockdevice_nodejs)

However, my IDE showed this message when hovering over the property:
```
property) NodeGroupBaseOptions.encryptRootBlockDevice?: Input<boolean> | undefined
Encrypt the root block device of the nodes in the node group.

@deprecated
This option has been deprecated for parameter naming coherence. Use the nodeRootVolumeEncrypted option instead.
```

Although the code accepts the `encryptRootBlockDevice` property, it is ignored.
One must set the undocumented `nodeRootVolumeEncrypted` property for this to work.

### Example

fwiw: https://gist.github.com/MitchellGerdisch/35e2c1ecee3f47dc01bf7f72b16213c8

### Output of `pulumi about`

CLI          
Version      3.89.0
Go Version   go1.21.1
Go Compiler  gc

Plugins
NAME        VERSION
aws         5.42.0
aws         5.31.0
awsx        1.0.6
docker      3.6.1
eks         1.0.4
kubernetes  3.30.2
nodejs      unknown

Host     
OS       darwin
Version  13.5
Arch     x86_64

This project is written in nodejs: executable='/Users/mitch/.nvm/versions/node/v18.0.0/bin/node' version='v18.0.0'

Backend        
Name           pulumi.com

Token type     personal

Dependencies:
NAME            VERSION
@pulumi/aws     5.42.0
@pulumi/awsx    1.0.6
@pulumi/eks     1.0.4
@pulumi/pulumi  3.89.0
@types/node     16.18.59

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by thomas11**
This PR modernizes the dependencies of the Pulumi [Amazon EKS](https://www.pulumi.com/registry/packages/eks/) component. 

## Main upgrades

- Pulumi AWS provider from v5 to v6. Version v6 brings [many changes](https://www.pulumi.com/registry/packages/aws/how-to-guides/6-0-migration/), but most of them are only relevant to the internals of EKS.
- Pulumi Kubernetes provider v3 to v4. It also [brings many improvements](https://www.pulumi.com/blog/kubernetes-4-0-even-more-kubernetes-native/#:~:text=Pulumi%20Kubernetes%204.0%20uses%20Kubernetes,resource%20type%20in%20the%20provider.) but they only affect the EKS internals.
- Pulumi Java language support from v0.6 to v0.9.7 for an improved Java SDK.
- The examples in this repository were upgraded to the latest dependencies including Go 1.21.

## Other changes

The deprecated property `Cluster.encryptRootBlockDevice` is removed. Resolves #926 and #930.

## Upgrading

At this point, no changes are expected in programs using pulumi-eks.

However, if the program also uses pulumi-aws, it will need to be upgraded to pulumi-aws v6, which has some [breaking changes](https://www.pulumi.com/registry/packages/aws/how-to-guides/6-0-migration/).

Note that as long as #929 is open, it's unlikely but possible that the upgrade tests might uncover more breaking changes.

Resolves #908 
Resolves #902
Resolves #930

### CODE REVIEW

File: nodejs/eks/cluster.ts [hunk 1]
```diff
@@ -313,8 +313,8 @@ export function getRoleProvider(
     parent?: pulumi.ComponentResource,
     provider?: pulumi.ProviderResource,
 ): CreationRoleProvider {
-    const partition = pulumi.output(aws.getPartition({ parent })).partition;
-    const accountId = pulumi.output(aws.getCallerIdentity({ parent })).accountId;
+    const partition = pulumi.output(aws.getPartition({})).partition;
+    const accountId = pulumi.output(aws.getCallerIdentity({})).accountId;
```
Comment (iwahbe) [inline nodejs/eks/cluster.ts:317]: Why are we no longer passing parent? Not passing parent causes `getRoleProvider` to fail when default providers are disabled.

File: nodejs/eks/nodegroup.ts [hunk 2]
```diff
@@ -1203,15 +1203,18 @@ ${customUserData}
...
+                owners: ["self", "amazon"],
+                filters: [
+                    {
+                        name: "image-id",
+                        values: [id],
+                    },
+                ],
+            },
+            { parent },
+        ),
```
Comment (iwahbe) [inline nodejs/eks/nodegroup.ts:1217]: Nit: formatting
```suggestion
        aws.ec2.getAmi({
           owners: ["self", "amazon"],
           filters: [
                {
                    name: "image-id",
                    values: [id],
                },
            ],
        }, { parent }),
```

This is just easier for me to read.

File: provider/go.mod [hunk 3]
```diff
@@ -1,9 +1,211 @@
 module github.com/pulumi/pulumi-eks/provider
 
-go 1.16
+go 1.20
```
Comment (iwahbe) [inline provider/go.mod:3]: If we're going to bump this, why not target go 1.21?

File: nodejs/eks/cluster.ts [hunk 4]
```diff
@@ -313,8 +313,8 @@ export function getRoleProvider(
     parent?: pulumi.ComponentResource,
     provider?: pulumi.ProviderResource,
 ): CreationRoleProvider {
-    const partition = pulumi.output(aws.getPartition({ parent })).partition;
-    const accountId = pulumi.output(aws.getCallerIdentity({ parent })).accountId;
+    const partition = pulumi.output(aws.getPartition({})).partition;
```
Comment (iwahbe) [inline nodejs/eks/cluster.ts:316]: Can we just use [`aws.getPartitionOutput`](https://www.pulumi.com/registry/packages/aws/api-docs/getpartition/) here?
```suggestion
    const partition = aws.getPartitionOutput({ parent }).partition;
```

**Review by iwahbe — COMMENTED** [review]
My understanding is that `nodejs/eks` is hand authored while `sdk/` is generated, so I haven't reviewed `sdk/`. For examples, I did a sanity check, but I can't actually confirm that they work as intended. I'm assuming that tests do that.

File: examples/aws-profile-role/package.json [hunk 5]
```diff
@@ -1,14 +1,13 @@
 {
     "name": "example-aws-profile-role",
     "devDependencies": {
-        "typescript": "^3.0.0",
+        "typescript": "^4.0.0",
         "@types/node": "latest"
     },
     "dependencies": {
-        "@pulumi/pulumi": "^3.0.0",
```
Comment (iwahbe) [inline examples/aws-profile-role/package.json:8]: We no longer have `@pulumi/pulumi` as a dependency here?

**Review by iwahbe — APPROVED** [review]
I don't see any problems. 

Make sure we don't merge until https://github.com/pulumi/pulumi-eks/actions/runs/6549739934 finishes.

**Comment by joeduffy** [comment]
@thomas11 @iwahbe What will the end user experience be upgrading to this package? I assume no downtime or replacements of existing resources? I didn't see upgrade tests but may have missed them.

**Comment by EvanBoyle** [comment]
Would be great to capture the answer to @joeduffy's question in the PR description. As a consumer of a project, I often look at commit messages (which usually default to PR descriptions) to understand what changed, why, and if there is any action I need to take.

None of that is clear from this PR.

File: examples/cluster-go/go.mod [hunk 6]
```diff
@@ -1,8 +1,87 @@
 module github.com/pulumi/pulumi-eks/examples/cluster-go
 
-go 1.16
+go 1.21
 
 require (
-	github.com/pulumi/pulumi-eks/sdk v0.40.0
-	github.com/pulumi/pulumi/sdk/v3 v3.36.0
+	github.com/pulumi/pulumi-eks/sdk/v2 v2.0.0-alpha.3
```
Comment (t0yv0) [inline examples/cluster-go/go.mod:6]: Nice!

**Pull request merged**